Add debug logging

Darrellbor opened the pull request:
PR for #70 

Implemented 
- [x] log write functionality that pushes `LogMessage` to the `loggerCh` channel
- [x] Add log write to all non-real-time methods of context
- [x] Add log write to all non-real-time methods of receiver
- [x] Add log write to all non-real-time methods of sender

Darrellbor commented:
@roc-streaming/go ready for review

gavv commented:
#74 is merged. Could you please **rebase** on fresh main?

Darrellbor commented:
> #74 is merged. Could you please **rebase** on fresh main?

done

gavv commented:
> > #74 is merged. Could you please **rebase** on fresh main?
> 
> done

You actually did not rebase, you did a merge. Commits from another pr are still included in this one.

Darrellbor commented:
> You actually did not rebase, you did a merge. Commits from another pr are still included in this one.
You are right, I thought you meant merge, I've rebased it

Darrellbor commented:
@roc-streaming/go ready for review

gavv commented on roc/context.go:
```diff
@@ -41,6 +42,11 @@ type Context struct {
 // Allocates and initializes a new context. May start some background threads.
 // User is responsible to call Context.Close to free context resources.
 func OpenContext(config ContextConfig) (*Context, error) {
+	logWrite(LogDebug, fmt.Sprintf("%v", logFormat{op: "roc_context_open_begin()",
+		params: map[string]interface{}{
+			"config":  config,
+			"version": Version(),
+		}}))
```
(3) `roc_context_open_begin` looks like an function name, but there is no such function, and it may be confusing. I suggest just to use current Go function everywhere to keep it simple:

```go
logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
```

gavv replied:
(4) `OpenContext` has multiple return paths, so just adding `logWrite` to the end is incomplete. Instead, we need to use defer:

```go
logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
defer logWrite(LogDebug, "leaving OpenContext()")
```

gavv replied:
(5) It would be very convenient to log returned error, if it's present. We could use named returns:

```go
func OpenContext(config ContextConfig) (ctx *Context, err error) {
    logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
    defer logWrite(LogDebug, "leaving OpenContext(): err=%v", err)
    ...
```

gavv replied:
(6) Finally, it would be convenient to log context / sender / receiver pointer when they're created, closed, and in all methods. This will allow distinguishing multiple instances of same type in logs.

```go
func OpenContext(config ContextConfig) (ctx *Context, err error) {
    logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
    defer logWrite(LogDebug, "leaving OpenContext(): context=%p err=%v", ctx, err)
    ...
```

```go
func OpenSender(context *Context, config SenderConfig) (sender *Sender, err error) {
    logWrite(LogDebug, "entering OpenSender(): context=%p config=%+v", context, config)
    defer logWrite(LogDebug, "leaving OpenSender(): context=%p sender=%p err=%v", ctx, sender, err)
    ...
```

gavv commented on roc/log.go:
```diff
@@ -181,6 +192,36 @@ func loggerRoutine() {
 	}
 }
 
+func getGID() uint64 {
+	b := make([]byte, 64)
+	b = b[:runtime.Stack(b, false)]
+	b = bytes.TrimPrefix(b, []byte("goroutine "))
+	b = b[:bytes.IndexByte(b, ' ')]
+	n, _ := strconv.ParseUint(string(b), 10, 64)
+	return n
+}
```
LogMessage TID field is for thread ID, not goroutine ID. I think it's better to keep filling it with TID, not GID, for two reasons:

* to avoid confusion when reading logs, otherwise you don't know who to interpret the integer you see if you don't know if the message originated from C or Go

* to keep track of native threads; we may see in logs what native threads are used in C code, and it may be important to understand if they are the same or different native threads used in Go code (e.g. when a Go callback is invoked from C)

Darrellbor replied:
I get errors flagged for this 
```
./log.go:6:1: error: unknown type name 'uint64_t'
uint64_t rocGoThreadID();
```

```
go list failed to return CompiledGoFiles. This may indicate failure to perform cgo processing; try building at the command line. See https://golang.org/issue/38990.go list
```

gavv replied:
This is just a prototype, I didn't test the code.

Darrellbor replied:
i'm not quite sure how to go about this

gavv replied:
No worries, then let's keep TID unset for now, and I'll implement the C part by myself after merging this PR.

Darrellbor replied:
alright noted

gavv commented on roc/log.go:
```diff
@@ -181,6 +192,36 @@ func loggerRoutine() {
+}
+
+func logWrite(level LogLevel, text string) {
+	if level >= LogDebug {
```
We should check `level` against currently used log level, not just `LogDebug`. For that, we'll need to remember level in `SetLogLevel()`. To avoid data race, we can use atomic.

Darrellbor replied:
Can you elaborate a bit more on this

gavv replied:
Your logErite func does tge following check: if given level is higher than debug, then print the message. 

Instead, the check should be: if given level is not higher than the level configured by user via SetLogLevel, then print it.

When the user invokes e.g. SetLogLevel(info), then only error and info messages should be printed, and debug and trace should be dropped.

See comment for SetLogLevel.

gavv commented on roc/log.go:
```diff
@@ -181,6 +192,36 @@ func loggerRoutine() {
+
+func logWrite(level LogLevel, text string) {
+	if level >= LogDebug {
+		module, ok := debug.ReadBuildInfo()
```
I think we can just hard-code module to "roc_go". It will fit well with other modules used for logging - big internal parts of roc-toolkit, like "roc_core", "roc_audio", "roc_rtp", etc.

gavv commented on roc/log.go:
```diff
@@ -181,6 +192,36 @@ func loggerRoutine() {
+func logWrite(level LogLevel, text string) {
+	if level >= LogDebug {
+		module, ok := debug.ReadBuildInfo()
+		_, file, line, callerOk := runtime.Caller(0)
+
+		if ok && callerOk {
```
I think we can log message even if callerOk is false. Missing file and line is not that critical.

Darrellbor replied:
test done

gavv reviewed (commented):
Thanks for PR!

gavv commented:
Note: don't forget to add version info logging to `versionCheck()`.

gavv commented:
Sorry for creating conflicts with version checking code.

gavv commented on roc/context.go:
```diff
@@ -40,7 +40,10 @@ type Context struct {
 // User is responsible to call Context.Close to free context resources.
-func OpenContext(config ContextConfig) (*Context, error) {
+func OpenContext(config ContextConfig) (ctx *Context, err error) {
+	logWrite(LogDebug, "entering roc_context_open(): config=%+v", config)
+	defer logWrite(LogDebug, "leaving roc_context_open(): context=%p err=%v", ctx, err)
```
Since these log messages surround theb whole OpenContex, not just the call to native roc_open_context, I suggest to replace roc_open_context with OpenContext in the message.

Same for other methods. Since we're logging the whole go method, let's use go method name.

Darrellbor replied:
done

gavv commented on roc/receiver.go:
```diff
@@ -171,7 +171,12 @@ type Receiver struct {
 // Allocates and initializes a new receiver, and attaches it to the context.
-func OpenReceiver(context *Context, config ReceiverConfig) (*Receiver, error) {
+func OpenReceiver(context *Context, config ReceiverConfig) (receiver *Receiver, err error) {
+	logWrite(LogDebug, "entering OpenReceiver(): config=%+v context=%p", config, context)
```
nit: everywhere we log context, let's make it the first in the log message, for consistency.

gavv commented on roc/receiver.go:
```diff
@@ -237,7 +242,12 @@ func OpenReceiver(context *Context, config ReceiverConfig) (*Receiver, error) {
 // Automatically initializes slot with given index if it's used first time.
-func (r *Receiver) SetMulticastGroup(slot Slot, iface Interface, ip string) error {
+func (r *Receiver) SetMulticastGroup(slot Slot, iface Interface, ip string) (err error) {
+	logWrite(LogDebug,
+		"entering Receiver.SetMulticastGroup(): slot=%v iface=%v ip=%v", slot, iface, ip,
+	)
```
nit: missing receiver/sender %p in "entering" message here and in many other places.

gavv commented on roc/receiver.go:
```diff
@@ -171,7 +171,12 @@ type Receiver struct {
 // Open a new receiver.
 // Allocates and initializes a new receiver, and attaches it to the context.
-func OpenReceiver(context *Context, config ReceiverConfig) (*Receiver, error) {
+func OpenReceiver(context *Context, config ReceiverConfig) (receiver *Receiver, err error) {
```
Here and in most other places, you introduced named returns, but did not modify the code to use them. In result, defer logs variables that were never set. Please test your code when you're submitting PR for review.

gavv commented on roc/version.go:
```diff
@@ -84,8 +84,12 @@ func (vi VersionInfo) Validate() error {
 }
 
 func checkVersion() {
+	var vi VersionInfo
+	logWrite(LogDebug, "entering versionCheck()")
+	defer logWrite(LogDebug, "leaving versionCheck(): version=%+v", vi)
```
There is no need to log every checkVersion call because in all invocations except first it's no-op.

Instead, it would be enough to log the retrieved version only once, after fetchVersion call and before validating it.

gavv commented on roc/log.go:
```diff
@@ -147,6 +149,7 @@ func (standardLogger) Print(v ...interface{}) {
 }
 
 var (
+	logLevel   atomic.Value
```
For consistency with other vars, let's rename it to `loggerLevel`.

There is no need to use atomic.Value here because it's just an int. You can use an int of an appropriate size instead, and corresponding atomic functions.

gavv commented on roc/log.go:
```diff
@@ -182,6 +185,24 @@ func loggerRoutine() {
+				Pid:   uint64(os.Getpid()),
+				//Tid:    -get from C module-
+				Module: "roc_go",
+				File:   "log.go",
+				Line:   210,
```
I guess you hard-coded file and line accidentally? Obtaining them from stack, as it was in your previous version of the patch, was a good implementation.

gavv commented on roc/log.go:
```diff
@@ -182,6 +185,24 @@ func loggerRoutine() {
+				Module: "roc_go",
+				File:   "log.go",
+				Line:   210,
+				Text:   fmt.Sprintf("%+v", text),
```
Did you test it? You're not properly formatting message text.

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +140,40 @@ func TestLog_Func(t *testing.T) {
 	}
 }
 
+func TestLog__LogWrite(t *testing.T) {
```
Extra `_`

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +140,40 @@ func TestLog_Func(t *testing.T) {
+
+	select {
+	case msg := <-ch:
+		assert.Equal(t, LogDebug, msg.Level, "Expected log level to be debug")
+		assert.NotEmpty(t, msg.Module)
+		assert.Equal(t, "roc_go", msg.Module)
+		assert.NotEmpty(t, msg.File)
+		assert.Equal(t, "log.go", msg.File)
+		assert.NotEmpty(t, msg.Line)
+		assert.Equal(t, 210, msg.Line)
```
nit: Could you also please check time, pid, tid, text, as I described earlier?

Darrellbor replied:
I added tests for all these except the exact value for Tid since it's gotten from the C module and I can't use that within the testing area, so instead I tested for nullability

gavv commented:
I've created PR for retrieving TID: https://github.com/roc-streaming/roc-go/pull/93

gavv commented on roc/log.c:
```diff
@@ -3,4 +3,4 @@
 void rocGoLogHandlerProxy(const roc_log_message* message, void* argument) {
     rocGoLogHandler((roc_log_message*)message);
-}
+}
```
Please rollback this change.

Darrellbor replied:
what change?

gavv replied:
Huh, github does not show it in all views. If you open https://github.com/roc-streaming/roc-go/pull/80/files, you can see it:

![image](https://github.com/roc-streaming/roc-go/assets/8436629/5872725c-edcd-421b-9fa1-b8c961495684)

It means that `\n` was removed from the end of the file, in other words, the last line in the file now does not have trailing newline, unlike all other lines. Probably your editor did that.

Darrellbor replied:
> Sorry for asking this again, but did you self-review and test your code before submitting it? I mean, enable logging and actually verify the logger output?
> 
> Formatting is still broken; the messages produced by logWrite are a mess.
> 
> This is a bit of wasting time of both of us, because I need to debug it locally, write a comment, then you need to read my comment, and debug it locally. So we're doubling debugging work and spend extra time for communication.

I'm sorry for this, but can i ask how i can enable logging and actually test it on my end... If there's a specific way that you do this

gavv commented:
> I've created PR for retrieving TID: #93

Merged. Could you please rebase and use new function? (`C.rocGoThreadID`) to fill TID?

gavv commented on roc/receiver.go:
```diff
@@ -171,18 +171,25 @@ type Receiver struct {
 // Allocates and initializes a new receiver, and attaches it to the context.
-func OpenReceiver(context *Context, config ReceiverConfig) (*Receiver, error) {
+func OpenReceiver(context *Context, config ReceiverConfig) (receiver *Receiver, err error) {
+	logWrite(LogDebug, "entering OpenReceiver(): context=%+v config=%p", context, config)
+	defer logWrite(LogDebug,
+		"leaving OpenReceiver(): context=%p receiver=%p err=%v", context, receiver, err,
+	)
```
In OpenSender and OpenReceiver, `%+v` and `%p` are messed up.

gavv commented on roc/receiver.go:
```diff
@@ -316,16 +341,23 @@ func (r *Receiver) SetReuseaddr(slot Slot, iface Interface, enabled bool) error
 // chosen ephemeral port. If the function succeeds, the actual port to which the
 // receiver was bound is written back to endpoint.
-func (r *Receiver) Bind(slot Slot, iface Interface, endpoint *Endpoint) error {
+func (r *Receiver) Bind(slot Slot, iface Interface, endpoint *Endpoint) (err error) {
+	logWrite(LogDebug,
+		"entering Receiver.Bind(): receiver=%p slot=%+v iface=%+v endpoint=%+v", r, slot, iface, endpoint,
+	)
+	defer logWrite(LogDebug, "leaving Receiver.Bind(): receiver=%p err=%v", r, err)
```
Since Bind() modifies endpoint argument, let's log endpoint in defer statement as well.

gavv commented on roc/version.go:
```diff
@@ -87,6 +87,9 @@ func checkVersion() {
 	if atomic.CompareAndSwapInt32(&versionCheckOnce, 0, 1) {
 		vi := fetchVersion()
 
+		logWrite(LogDebug, "entering versionCheck()")
+		defer logWrite(LogDebug, "leaving versionCheck(): version=%+v", vi)
```
As I already wrote before, there is no need for entering/leaving message pair here, we can just log version after fetching it and before validating it. There is no big work done between enter and leave, unlike other methods.

Also using defer hurts here, because this way Validate() will panic before version is logged.

gavv commented on roc/log.go:
```diff
@@ -183,6 +187,23 @@ func loggerRoutine() {
 }
 
+func logWrite(level LogLevel, text ...interface{}) {
+	if level <= loggerLevel {
```
This is a data race because you're reading loggerLevel in logWrite and writing it in SetLogLevel concurrently without protections.

As I already wrote, you can make loggerLevel a plain integer, but you still need to use appropriatre atomic operations to work with it.

gavv commented on roc/log.go:
```diff
@@ -183,6 +187,23 @@ func loggerRoutine() {
 
+func logWrite(level LogLevel, text ...interface{}) {
+	if level <= loggerLevel {
+		_, file, line, _ := runtime.Caller(0)
```
It seems we're running circles here.

As I already wrote, this will return file and line that corresponds to `log.go`. In other words, file and line will always be `log:go` and `192`.

What we need instead is to obtain the file and line which correspond to the place from which `logWrite()` was invoked.

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+	select {
+	case msg := <-ch:
+		assert.Equal(t, LogDebug, msg.Level, "Expected log level to be debug")
+		assert.NotEmpty(t, msg.Module)
+		assert.Equal(t, "roc_go", msg.Module)
```
NotEmpty for Module is redundant because we already have Equal check.

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+		assert.Equal(t, LogDebug, msg.Level, "Expected log level to be debug")
+		assert.NotEmpty(t, msg.Module)
+		assert.Equal(t, "roc_go", msg.Module)
+		assert.NotEmpty(t, msg.File)
+		assert.Equal(t, "log.go", msg.File[len(msg.File)-6:])
```
After you fix file & line filling, `File` should become "context.go" in this test. NotEmpty is redundant.

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+		assert.Equal(t, "roc_go", msg.Module)
+		assert.NotEmpty(t, msg.File)
+		assert.Equal(t, "log.go", msg.File[len(msg.File)-6:])
+		assert.NotEmpty(t, msg.Line)
+		assert.Equal(t, 192, msg.Line)
```
Hard-coding specific line is not very good, it may frequently change due to modifications not related to this test, and it should not break the test.

Instead of NotEmpty + Equal, let's just check that line is positive number.

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+		assert.Equal(t, "log.go", msg.File[len(msg.File)-6:])
+		assert.NotEmpty(t, msg.Line)
+		assert.Equal(t, 192, msg.Line)
+		assert.NotEmpty(t, msg.Pid)
+		assert.Equal(t, uint64(os.Getpid()), msg.Pid)
```
NotEmpty check is redundant here

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+		assert.NotEmpty(t, msg.Pid)
+		assert.Equal(t, uint64(os.Getpid()), msg.Pid)
+		assert.NotEmpty(t, msg.Tid)
+		assert.NotEmpty(t, msg.Time)
+		assert.Equal(t, time.Now().Truncate(time.Minute), msg.Time.Truncate(time.Minute))
```
* NotEmpty is redundant
* A more natural and correct check would be to remember time at the start of the test and check that the message time is not before that remembered time and not after now

gavv commented on roc/log_test.go:
```diff
@@ -140,6 +141,47 @@ func TestLog_Func(t *testing.T) {
+		assert.NotEmpty(t, msg.Tid)
+		assert.NotEmpty(t, msg.Time)
+		assert.Equal(t, time.Now().Truncate(time.Minute), msg.Time.Truncate(time.Minute))
+		assert.NotEmpty(t, msg.Text)
+		assert.Equal(t, "entering OpenContext(): config=%+v%!(EXTRA roc.ContextConfig={0 0})", msg.Text)
```
* NotEmpty is redundant
* I think equality check is too strict here, we can just do some sanity check, e.g. ensure that text contains "OpenContext" or maybe "entering OpenContext"

gavv commented:
Sorry for asking this again, but did you self-review and test your code before submitting it? I mean, enable logging and actually verify the logger output?

Formatting is still broken; the messages produced by logWrite are a mess.

This is a bit of wasting time of both of us, because I need to debug it locally, write a comment, then you need to read my comment, and debug it locally. So we're doubling debugging work and spend extra time for communication.

gavv commented:
Regarding named returns. The current code looks a bit messy: it mixes named returns with regular returns like this:

```go
	if endpoint == nil {
		err = errors.New("endpoint is nil")
		return err
	}
```

In other words, we first fill named return var `err`, but then anyway use regular `return err`.

To keep code easier to follow, I suggest to completely switch to **naked** returns everywhere where we use logWrite, like this:

```go
	if endpoint == nil {
		err = errors.New("endpoint is nil")
		return
	}
```

and:

```go
err = nil
return
```

gavv commented:
We forgot to update 1 place to use named return vars:
https://github.com/roc-streaming/roc-go/blob/61892242983ae2ca830c6ff743160071569ab8a6/roc/context.go#L56C19-L58

gavv commented:
For consistency, I suggest to update ReadFloats and WriteFloats to use named return vars + naked returns too.

Darrellbor commented:
> Sorry for asking this again, but did you self-review and test your code before submitting it? I mean, enable logging and actually verify the logger output?
> 
> Formatting is still broken; the messages produced by logWrite are a mess.
> 
> This is a bit of wasting time of both of us, because I need to debug it locally, write a comment, then you need to read my comment, and debug it locally. So we're doubling debugging work and spend extra time for communication.

I'm sorry for this, but can i ask how i can enable logging and actually test it on my end... If there's a specific way that you do this

gavv commented:
You can just hack log.go and maybe main_test.go to ensure logs are not disabled in tests.

gavv commented:
> To keep code easier to follow, I suggest to completely switch to **naked** returns everywhere where we use logWrite, like this:
> 
> ```go
> 	if endpoint == nil {
> 		err = errors.New("endpoint is nil")
> 		return
> 	}
> ```

I have created a small linter that will help us to maintain this code style: https://github.com/gavv/returnstyles

You can install it to GOPATH using:

```
go install github.com/gavv/returnstyles@latest
```

And then run like this:

```
cd roc
returnstyles -named-allow=true -named-allow-normal-returns=false .
```

(see linter README for explanations)

It will report messages like:

```
/home/victor/dev/roc-streaming/roc-go/roc/context.go:57:3: normal (non-naked) returns not allowed in functions with named return variables
/home/victor/dev/roc-streaming/roc-go/roc/context.go:67:2: normal (non-naked) returns not allowed in functions with named return variables
/home/victor/dev/roc-streaming/roc-go/roc/context.go:85:4: normal (non-naked) returns not allowed in functions with named return variables
/home/victor/dev/roc-streaming/roc-go/roc/context.go:91:2: normal (non-naked) returns not allowed in functions with named return variables
```

It should help you to find all problematic places that should be fixed. Later we'll integrate this linter into makefile and CI.

Darrellbor commented:
@gavv ready for re-review

gavv commented:
```
2023/05/22 22:28:34 [dbg] roc_go: entering OpenContext(): config={MaxPacketSize:0 MaxFrameSize:0}
...
2023/05/22 22:28:34 [dbg] roc_go: leaving OpenContext(): context=0x0 err=<nil>
```

We have a problem - context is 0x0 in "leaving" message. See explanation here: https://stackoverflow.com/questions/42703707/

To fix it, we can convert this:

```go
	logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
	defer logWrite(LogDebug, "leaving OpenContext(): context=%p err=%v", ctx, err)
```

to this:

```go
	logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
	defer func() {
		logWrite(LogDebug, "leaving OpenContext(): context=%p err=%v", ctx, err)
	}()
```

Could you please apply this fix everywhere and test that everything is logged correctly?

gavv commented:
Everywhere where we log errors, I suggest to use `err=%#v` instead of `err=%v`.

With `%v`:

```
2023/05/22 23:06:25 [dbg] roc_go: leaving Receiver.SetMulticastGroup():
  receiver=0xc00001e6a0 err=roc_receiver_set_multicast_group() failed with code -1
```

With `%#v`:

```
2023/05/22 23:06:47 [dbg] roc_go: leaving Receiver.SetMulticastGroup():
  receiver=0xc000172500 err=roc.nativeErr{op:"roc_receiver_set_multicast_group()", code:-1}
```

These logs are intended to be read by bindings developers, and I think for this purpose `%#v` is more informative.

gavv commented:
There is a difference how File field is formatted by bindings and native lib:

```
2023/05/22 23:30:24 [dbg] roc_go: /home/victor/dev/roc-streaming/roc-go/roc/context.go:44
  entering OpenContext(): config={MaxPacketSize:0 MaxFrameSize:0}
```

```
2023/05/22 23:30:24 [dbg] roc_core: src/internal_modules/roc_core/slab_pool.cpp:38
  slab pool: initializing: object_size=672 min_slab=0B(1S) max_slab=0B(0S) poison=0
```

For bindings, it's an absolute path: `/home/victor/dev/roc-streaming/roc-go/roc/version.go`

For native lib, it's a path relative to project root: `src/internal_modules/roc_core/slab_pool.cpp`

It would be nice to unify formatting and make bindings path relative to project root too, i.e. instead of `/home/victor/dev/roc-streaming/roc-go/roc/version.go` show `roc/version.go`.

We can achieve this by finding the most right `/roc/` component in the path and truncating everything that is on the left of it.

Note: we should do it in a cross-platform way, i.e. we should not rely on specific path separator (`/` or `\`).

gavv commented:
(I think we should do it in logWrite)

gavv commented:
If this commit: https://github.com/roc-streaming/roc-go/pull/80/commits/4fef44facb574d98aae4f85061534a02873fa8fa

Is ready for review and is intended to fix this: https://github.com/roc-streaming/roc-go/pull/80#issuecomment-1557835660

Then it's not the algorithm I suggested. This implementation makes an assumption about folder nesting and will break if it's changed somehow (e.g. we add subdirs).

I suggested a more relaxed requirement which will work with any nesting.

Other ideas are welcome too.

gavv commented:
PR for adding `returnstyles` linter to repo: https://github.com/roc-streaming/roc-go/pull/95

Darrellbor commented:
> If this commit: [4fef44f](https://github.com/roc-streaming/roc-go/commit/4fef44facb574d98aae4f85061534a02873fa8fa)
> 
> Is ready for review and is intended to fix this: [#80 (comment)](https://github.com/roc-streaming/roc-go/pull/80#issuecomment-1557835660)
> 
> Then it's not the algorithm I suggested. This implementation makes an assumption about folder nesting and will break if it's changed somehow (e.g. we add subdirs).
> 
> I suggested a more relaxed requirement which will work with any nesting.
> 
> Other ideas are welcome too.

This was one way of doing it, alternatively i thought of splitting by `\` and `/` and joining the last two together but felt the path route would be platform agnostic.

Darrellbor commented on roc/log_test.go:
```diff
@@ -140,6 +141,49 @@ func TestLog_Func(t *testing.T) {
+			assert.True(t,
+				msg.Time.After(testStartTime),
+				"Time assertion failed: test time is less than start time of the test",
+			)
```
I introduced a tolerance of a millisecond to compare the time to also account for when the time is the same

Darrellbor commented on roc/log_test.go:
```diff
@@ -140,6 +141,49 @@ func TestLog_Func(t *testing.T) {
+
+	select {
+	case msg := <-ch:
+		if msg.Module == "roc_go" && strings.Contains(msg.Text, "entering OpenContext()") {
```
I tried removing this but I was getting `leaving Context.Close()` being pushed into the channel instead

gavv commented:
> This was one way of doing it, alternatively i thought of splitting by \ and / and joining the last two together but felt the path route would be platform agnostic.

Joining **the last two** will lead to the same problem: we're hard-coding fixed nesting level (2); I'd like to avoid it, it's very implicit and not very flexible.

gavv commented:
You could use

gavv commented on roc/log_test.go:
```diff
@@ -180,3 +181,51 @@ func TestLog_Levels(t *testing.T) {
+		assert.Greater(t, msg.Line, 0, "Line number must be positive")
+		assert.Equal(t, uint64(os.Getpid()), msg.Pid)
+		assert.NotEmpty(t, msg.Tid)
+		assert.True(t,
+			msg.Time.After(testStartTime.Add(-time.Millisecond)) &&
+				msg.Time.Before(testStartTime.Add(time.Millisecond)),
+			"Time assertion failed: test time is not within the tolerance of the start time of the test",
+		)
+		assert.True(t,
+			msg.Time.After(time.Now().Add(-time.Millisecond)) &&
+				msg.Time.Before(time.Now().Add(time.Millisecond)),
+			"Time assertion failed: message time cannot be greater than the time now",
+		)
```
```suggestion
		assert.WithinRange(t, msg.Time, testStartTime, time.Now(),
			"Time must have meaningful value")
```

gavv replied:
This way there is no need for magic millisecond because the range is inclusive. Also the code is much shorter.

Darrellbor replied:
WithinRange doesn't seem to exist on the assert package

Darrellbor replied:
assert.WithinRange doesn't exist though

gavv commented:
I've created issue to allow enabling logs in tests: https://github.com/roc-streaming/roc-go/issues/96

Darrellbor commented:
> You could use https://pkg.go.dev/path/filepath#SplitList

I attempted using this but it just returns the path in full
I use it after using the `filepath.ToSlash(file)` method

gavv commented:
A guy on reddit [suggested](https://www.reddit.com/r/golang/comments/13njcs0/comment/jl9jl5a/?utm_source=share&utm_medium=web2x&context=3) that naked returns are not actually needed here.

Earlier I did an incorrect test and decided that we need naked returns. But in fact, if we properly wrap deferred code in `func(){}`, as we already did recently, we can still use normal (non-naked) returns and return variables will have desired values in defer:

```go
	logWrite(LogDebug, "entering OpenContext(): config=%+v", config)
	defer func() {
		logWrite(LogDebug, "leaving OpenContext(): context=%p err=%#v", ctx, err)
	}()

	...

	if errCode != 0 {
		return nil, newNativeErr("roc_context_open()", errCode)
	}
```

The code is cleaner with non-naked returns, so could we please switch back to using them?

Sorry for the mess!

gavv commented:
> > You could use https://pkg.go.dev/path/filepath#SplitList
> 
> I attempted using this but it just returns the path in full I use it after using the `filepath.ToSlash(file)` method

Oh, you're right, SplitList is for splitting colon-separated list of paths (like `PATH`), not for splitting single path into components.

Well then I guess you could use filepath.Split iteratively until you reach "roc" parent directory, or something like this.

Darrellbor commented:
```
if errCode != 0 {
		return nil, newNativeErr("roc_context_open()", errCode)
	}
```
Wouldn't i need to still assign the values of `ctx` and `err`?

gavv commented:
> ```
> if errCode != 0 {
> 		return nil, newNativeErr("roc_context_open()", errCode)
> 	}
> ```
> 
> Wouldn't i need to still assign the values of `ctx` and `err`?

I also suggest to expand logWrite test a bit, and check not only "enetering" but also "leaving" message and check that the leaving message contains "context=0x..." where "..." is not zero address.

Darrellbor commented:
@gavv 
I have worked on the following
- extended the test to accommodate `leaving OpenContext()` as well as checking that the `context=0x...` is accounted for
- reverted back to non-naked returns style for all the previously changed methods
- reverted back to the method I used in the first place as `filepath.Split` didn't do what I intended and would result in the same approach for implementation.
- assert.WithinRange doesn't exist so I reverted back to the assert.True methods

gavv commented on roc/log_test.go:
```diff
@@ -180,3 +182,56 @@ func TestLog_Levels(t *testing.T) {
+		}
+	}
+
+	ctx.Close()
```
nit: better use `defer ctx.Close()` after creating context.

gavv commented on roc/receiver.go:
```diff
@@ -237,17 +244,24 @@ func OpenReceiver(context *Context, config ReceiverConfig) (*Receiver, error) {
+	logWrite(LogDebug,
+		"entering Receiver.SetMulticastGroup(): receiver=%p slot=%v iface=%v ip=%v", r, slot, iface, ip,
+	)
+	defer func() {
+		logWrite(LogDebug, "leaving Receiver.SetMulticastGroup(): receiver=%p err=%#v", r, err)
+	}()
+
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 
 	if r.cPtr == nil {
 		return errors.New("receiver is closed")
 	}
 
-	cIP, err := go2cStr(ip)
-	if err != nil {
-		return fmt.Errorf("invalid ip: %w", err)
+	cIP, parseErr := go2cStr(ip)
+	if parseErr != nil {
+		return fmt.Errorf("invalid ip: %w", parseErr)
 	}
```
nit: parseErr can be renamed back to err.

gavv commented on roc/receiver.go:
```diff
@@ -377,7 +409,7 @@ func (r *Receiver) Bind(slot Slot, iface Interface, endpoint *Endpoint) error {
 //
 // Until the receiver is connected to at least one sender, it produces silence.
 // If the receiver is connected to multiple senders, it mixes their streams into one.
-func (r *Receiver) ReadFloats(frame []float32) error {
+func (r *Receiver) ReadFloats(frame []float32) (err error) {
```
nit: named return can be reverted.

gavv commented on roc/log.go:
```diff
@@ -183,6 +189,40 @@ func loggerRoutine() {
+	var finalPath string
+	_, file, line, _ := runtime.Caller(1)
+
+	dirs := strings.FieldsFunc(file, func(c rune) bool {
+		return c == '/' || c == '\\'
+	})
+	fileNameIndex := len(dirs) - 1
+	lastDirIndex := len(dirs) - 2
+
+	if lastDirIndex >= 0 {
+		fileName := dirs[fileNameIndex]
+		lastDir := dirs[lastDirIndex]
+
+		finalPath = filepath.Join(lastDir, fileName)
+	}
```
This is the same algorithm that you already implemented in previous versions, just expressed in different way. You're still implicitly hard-coding specific nesting level (2). Please read carefully what algorithm I suggested as an alternative and why, or feel free to suggest other algorithms that avoid hard-coding nesting level. https://github.com/roc-streaming/roc-go/pull/80#issuecomment-1557835660

gavv reviewed (commented):
Thanks for update

gavv commented on roc/log.go:
```diff
@@ -183,6 +189,40 @@ func loggerRoutine() {
 	}
 }
 
+func logWrite(level LogLevel, text string, params ...interface{}) {
+	storedLogLevel := LogLevel(atomic.LoadInt32(&loggerLevel))
+	if level > storedLogLevel {
+		return
+	}
+
+	var finalPath string
+	_, file, line, _ := runtime.Caller(1)
+
+	dirs := strings.FieldsFunc(file, func(c rune) bool {
+		return c == '/' || c == '\\'
+	})
+	fileNameIndex := len(dirs) - 1
+	lastDirIndex := len(dirs) - 2
+
+	if lastDirIndex >= 0 {
+		fileName := dirs[fileNameIndex]
+		lastDir := dirs[lastDirIndex]
+
+		finalPath = filepath.Join(lastDir, fileName)
+	}
+
+	loggerCh <- LogMessage{
+		Level:  level,
+		Time:   time.Now(),
+		Pid:    uint64(os.Getpid()),
+		Tid:    uint64(C.rocGoThreadID()),
+		Module: "roc_go",
+		File:   finalPath,
+		Line:   line,
+		Text:   fmt.Sprintf(text, params...),
+	}
+}
```
```suggestion
func logWrite(level LogLevel, text string, params ...interface{}) {
	storedLogLevel := LogLevel(atomic.LoadInt32(&loggerLevel))
	if level > storedLogLevel {
		return
	}

	file, line := logLocation(2)

	loggerCh <- LogMessage{
		Level:  level,
		Time:   time.Now(),
		Pid:    uint64(os.Getpid()),
		Tid:    uint64(C.rocGoThreadID()),
		Module: "roc_go",
		File:   file,
		Line:   line,
		Text:   fmt.Sprintf(text, params...),
	}
}

func logLocation(stack int) (string, int) {
	_, file, line, ok := runtime.Caller(stack)
	if !ok {
		return "", -1
	}

	parts := strings.FieldsFunc(file, func(c rune) bool {
		return c == '/' || c == '\\'
	})
	if len(parts) == 0 {
		return "", -1
	}

	for i := len(parts) - 1; i != 0; i-- {
		if parts[i] == "roc" {
			parts = parts[i:]
			break
		}
	}

	return filepath.Join(parts...), line
}
```

gavv replied:
Example implementation of the algorithm I was talking about.
